Added focus_on_window_activation directive

Airblader opened the pull request:
Before this gets merged, I want to add testcases for it. I just submitted the PR already because I'm not entirely sure about the naming. Opinions?

Airblader commented:
@stapelberg To make sure, we don't want to change the behavior for such requests coming from a pager, right?

Deiz commented on docs/userguide:
```diff
@@ -981,6 +981,31 @@ force_display_urgency_hint <timeout> ms
+
+*Syntax*:
+--------------------------------------------------
+focus_on_window_activation smart|urgent|focus|none
```
Small nitpick, the parameter lists (here and in config.spec) should probably be enclosed in &lt;&gt;, e.g. &lt;smart|urgent|focus|none&gt;

Airblader replied:
Yep, thanks. I've changed it.

acrisci commented:
This looks kind of ugly to me.

Airblader commented:
@acrisci I'll be happy to change it if you can be a bite more specific.

stapelberg commented on include/config.h:
```diff
@@ -167,6 +167,18 @@ struct Config {
+    /** Behavior when a window sends a NET_ACTIVE_WINDOW message. */
+    enum {
+        /* Focus if the target workspace is visible, set urgency hint otherwise. */
+        FOWN_SMART,
```
These should be FOWA, not FOWN, right?

stapelberg commented on docs/userguide:
```diff
@@ -981,6 +981,31 @@ force_display_urgency_hint <timeout> ms
 force_display_urgency_hint 500 ms
 ---------------------------------
 
+=== Behavior when a window is activated
```
I’d rename this to “Focus on window activation” to make it clearer what it’s talking about and consistent with how the option itself is called.

stapelberg commented:
Looks good overall.

Airblader committed:
Updated userguide for the directive focus_on_window_activation introduced in #1426.

Airblader committed:
Added 'focus_on_window_activation' directive

When a window receives a _NET_ACTIVE_WINDOW message, it can steal the focus. This may not be preferable to all users.
With this directive, the user can choose from one of the following:
1) 'smart' - focus the container if its workspace is visible, otherwise set the urgency flag (default)
2) 'urgent' - always set the urgency flag, do not steal focus
3) 'focus' - always switch focus, never set the urgency hint
4) 'none' - ignore the request entirely (do not switch focus, nor set the urgency hint)

fixes #1426

Airblader committed:
Added testcases for #1426

Airblader commented:
@stapelberg I've added testcases, incorporated your comments and rebased it.

Airblader commented on testcases/t/240-focus-on-window-activation.t:
```diff
@@ -0,0 +1,206 @@
+    return $current->{urgent};
+}
+
+#####################################################################
```
I'm not testing "smart" here since it's the default and extensively tested in another test. I had to create a new test class for these since they need to run with their own config.

stapelberg commented:
Nice!
